routes/index.js

```diff
@@ -1,142 +1,166 @@
     if (!password) {
-        res.render("error", { message: "Invalid password." });
+        res.status(403).json({
+            success: false,
+            message: "Invalid password",
+        });
         return;
     }
 
-    const account = game.createAccount(username, password);
+    game.createAccount(username, password);
     game.save();
-    req.session.accountId = account.id;
-    res.redirect("/");
+    res.json({ success: true, message: "Account created successfully" });
 });
 
 router.post("/login", (req, res) => {
     const username = req.body.username;
     const account = game.getAccountByUsername(username);
     if (!account) {
-        res.render("error", {
+        res.status(401).json({
+            success: false,
             message: `The username '${username}' doesn't exist, create an account first.`,
         });
         return;
     }
+
     const password = req.body.password;
     if (
         !Password.verify(password, account.passwordSalt, account.passwordHash)
     ) {
-        res.render("error", {
+        res.status(401).json({
+            success: false,
             message: "Wrong password.",
         });
         return;
     }
 
-    req.session.accountId = account.id;
-    res.redirect("/");
+    const token = generateToken({
+        ...account,
+        passwordHash: "you wish",
+        passwordSalt: "you wish",
+    });
+    res.json({
+        success: true,
+        message: "Authentication successful!",
+        token,
+        user: account,
+    });
 });
 
 router.get("/logout", (req, res) => {
     delete req.session.accountId;
+    delete req.user;
+    // TODO: Also do some invalidation of token.
     res.redirect("/");
 });
 
-router.get("/shader", (req, res) => {
-    const account = game.getAccountById(req.session.accountId);
-    if (!account) {
-        invalidAccountRender(res, req.session.accountId);
-        return;
-    }
-
-    res.render("shader", { account: account });
-});
-
 router.post("/shader", (req, res) => {
-    const account = game.getAccountById(req.session.accountId);
+    const account = game.getAccountById(req.user.id);
     if (!account) {
-        invalidAccountRender(res, req.session.accountId);
+        res.status(401).json({
+            success: false,
+            message: "Could not find any account tied to your token",
+        });
         return;
     }
-    account.vertexShaderCode = req.body.vertexShader;
-    account.fragmentShaderCode = req.body.fragmentShader;
 
-    account.useShaderForProfilePicture =
-        req.body.useShaderForProfilePicture === "on";
+    account.vertexShaderCode = req.body.vertexShaderCode;
+    account.fragmentShaderCode = req.body.fragmentShaderCode;
+    account.useShaderForProfilePicture = req.body.useShaderForProfilePicture;
 
     game.save();
-    res.redirect("/");
+    res.json({
+        success: true,
+        message: "Updated shader profile",
+        account: { ...account, passwordHash: "", passwordSalt: "" },
```

Comment:
can we middleware this thing somehow, or maybe we should split it so account is like
account.data = the stuff we have except password
account.cred = password stuff
and just send account.data?

wanna get rid of having to manipulate the accounts array each time we send all accounts though, any idea for separating it?

out of scope for this PR of course

Reply:
We should probably be able to do this in a nice way, maybe without having to change the underlying structure too much and without middleware. It feels like there should be a way to make the sensitive fields more of a optin when getting data.

Reply:
https://github.com/Pluppen/HealtyBoiChallenge/issues/17